add function getWindows()

funkhou9 opened the pull request:
This has been tested using a BGData object. Intended to also work with a BEDMatrix, although this hasn't been tested. No formal unit tests have been built for getWindows() yet.

Commit:
initial commit...

add function getWindows() for obtaining sliding LD-based windows given a genotype object such as BGData object.

Commit:
fix error...

error would be thrown unless user specified a range that was equal to the number of columns of X.

anainesvs commented:
<!--
/* Font Definitions */
@font-face
	{font-family:Wingdings;
	panose-1:5 0 0 0 0 0 0 0 0 0;}
@font-face
	{font-family:"Cambria Math";
	panose-1:2 4 5 3 5 4 6 3 2 4;}
@font-face
	{font-family:Calibri;
	panose-1:2 15 5 2 2 2 4 3 2 4;}
/* Style Definitions */
p.MsoNormal, li.MsoNormal, div.MsoNormal
	{margin:0in;
	margin-bottom:.0001pt;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
h4
	{mso-style-priority:9;
	mso-style-link:"Heading 4 Char";
	mso-margin-top-alt:auto;
	margin-right:0in;
	mso-margin-bottom-alt:auto;
	margin-left:0in;
	font-size:12.0pt;
	font-family:"Calibri",sans-serif;
	font-weight:bold;}
a:link, span.MsoHyperlink
	{mso-style-priority:99;
	color:blue;
	text-decoration:underline;}
a:visited, span.MsoHyperlinkFollowed
	{mso-style-priority:99;
	color:purple;
	text-decoration:underline;}
p.msonormal0, li.msonormal0, div.msonormal0
	{mso-style-name:msonormal;
	mso-margin-top-alt:auto;
	margin-right:0in;
	mso-margin-bottom-alt:auto;
	margin-left:0in;
	font-size:11.0pt;
	font-family:"Calibri",sans-serif;}
span.Heading4Char
	{mso-style-name:"Heading 4 Char";
	mso-style-priority:9;
	mso-style-link:"Heading 4";
	font-family:"Calibri Light",sans-serif;
	color:#2F5496;
	font-style:italic;}
span.EmailStyle21
	{mso-style-type:personal-reply;
	font-family:"Calibri",sans-serif;
	color:windowtext;}
.MsoChpDefault
	{mso-style-type:export-only;
	font-size:10.0pt;}
@page WordSection1
	{size:8.5in 11.0in;
	margin:1.0in 1.0in 1.0in 1.0in;}
div.WordSection1
	{page:WordSection1;}
/* List Definitions */
@list l0
	{mso-list-id:1092819801;
	mso-list-template-ids:-1705078292;}
@list l0:level1
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l0:level2
	{mso-level-number-format:bullet;
	mso-level-text:o;
	mso-level-tab-stop:1.0in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:"Courier New";
	mso-bidi-font-family:"Times New Roman";}
@list l0:level3
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:1.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l0:level4
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:2.0in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l0:level5
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:2.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l0:level6
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:3.0in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l0:level7
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:3.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l0:level8
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:4.0in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l0:level9
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:4.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l1
	{mso-list-id:1360467886;
	mso-list-template-ids:1300667916;}
@list l1:level1
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l1:level2
	{mso-level-number-format:bullet;
	mso-level-text:o;
	mso-level-tab-stop:1.0in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:"Courier New";
	mso-bidi-font-family:"Times New Roman";}
@list l1:level3
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:1.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l1:level4
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:2.0in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l1:level5
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:2.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l1:level6
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:3.0in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l1:level7
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:3.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l1:level8
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:4.0in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l1:level9
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:4.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l2
	{mso-list-id:1513759900;
	mso-list-template-ids:1814301050;}
@list l2:level1
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Symbol;}
@list l2:level2
	{mso-level-number-format:bullet;
	mso-level-text:o;
	mso-level-tab-stop:1.0in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:"Courier New";
	mso-bidi-font-family:"Times New Roman";}
@list l2:level3
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:1.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l2:level4
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:2.0in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l2:level5
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:2.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l2:level6
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:3.0in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l2:level7
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:3.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l2:level8
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:4.0in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
@list l2:level9
	{mso-level-number-format:bullet;
	mso-level-text:;
	mso-level-tab-stop:4.5in;
	mso-level-number-position:left;
	text-indent:-.25in;
	mso-ansi-font-size:10.0pt;
	font-family:Wingdings;}
ol
	{margin-bottom:0in;}
ul
	{margin-bottom:0in;}
-->



Thank you
 

From: Scott Funkhouser <notifications@github.com>
Reply-To: QuantGen/BGData <reply@reply.github.com>
Date: Monday, July 16, 2018 at 10:09 AM
To: QuantGen/BGData <BGData@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: [QuantGen/BGData] add function getWindows() (#50)


 

This has been tested using a BGData object. Intended to also work with a BEDMatrix, although this hasn't been tested. No formal unit tests have been built for getWindows() yet.



You can view, comment on, or merge this pull request online at:
  https://github.com/QuantGen/BGData/pull/50
Commit Summary


initial commit... 
fix error... 
File Changes


M 
NAMESPACE (1) 
A 
R/getWindows.R (99) 
A 
man/getWindows.Rd (27) 
Patch Links:


https://github.com/QuantGen/BGData/pull/50.patch

https://github.com/QuantGen/BGData/pull/50.diff

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, 
view it on GitHub, or 
mute the thread.

agrueneberg commented on R/getWindows.R:
```diff
@@ -0,0 +1,99 @@
+#' @export
+getWindows <- function(X, range = 1:100, rSq = 0.1, maxGaps = 2) {
+    Xwin <- X[, range]
+    win <- lapply(1:ncol(Xwin), function(i) getBlock(Xwin, i, rSq = r2, maxGaps = gaps))
```
`r2` should be `rSq`.

funkhou9 replied:
Oops yep thanks for spotting the `r2` and `gaps` mistakes!

agrueneberg commented on R/getWindows.R:
```diff
@@ -0,0 +1,99 @@
+#' @export
+getWindows <- function(X, range = 1:100, rSq = 0.1, maxGaps = 2) {
+    Xwin <- X[, range]
+    win <- lapply(1:ncol(Xwin), function(i) getBlock(Xwin, i, rSq = r2, maxGaps = gaps))
```
`gaps` should be `maxGaps`.

agrueneberg commented on R/getWindows.R:
```diff
@@ -0,0 +1,99 @@
+    lag=1
+    nGaps=0
+    while(!ready){
+        R2=cor(xi,X[,center-lag],use='complete.obs')^2
```
Please use `stats::cor` to make `R CMD check` happy.

agrueneberg commented on R/getWindows.R:
```diff
@@ -0,0 +1,99 @@
+    lag=1
+    nGaps=0
+    while(!ready){
+        R2=cor(xi,X[,center+lag],use='complete.obs')^2
```
Please use `stats::cor` to make `R CMD check` happy.

agrueneberg reviewed (changes requested):
Looks good to me! Thanks for the NAMESPACE entry and the documentation. I've requested a few minor changes.

Commit:
make requested changes

funkhou9 commented:
Beyond your requested changes, I'm still finding little improvements that I'm continuing to add. Those'll be pushed shortly.

Commit:
change argument and add new argument

arg range changed to center_snp. improves functionality by no longer attempting to subset SNPs from the BGData object.

added arg ids so that ld can be estimated using a subset of individuals in X.

Commit:
center snps now shown as indices rather than snp names

agrueneberg commented on R/getWindows.R:
```diff
@@ -0,0 +1,99 @@
+#' each element of `X`.
+#' @param rSq R-squared threshold for determining genotype correlations.
+#' @param maxGaps The number of consequtive SNPs that may fall below `rSq` until LD is considered lost
+#' @param ids Indices of individuals used to compute LD, if not all individuals in X are desired.
```
In other functions in BGData this parameter is just called `i`. It would be nice to keep it consistent. You can use `i <- crochet::convertIndex(X, i, "i")` at the beginning of the function to convert different types of indices to positive integers, e.g. when someone passes in a set of sample names.

agrueneberg replied:
It may also make sense to add a `j` parameter to exclude some variants that for example were removed by QC. Not sure how we should treat the range then, as a range of `X` before or after subsetting?

agrueneberg commented on R/getWindows.R:
```diff
@@ -0,0 +1,99 @@
+#'
+#' @param X A matrix-like object, typically `@@geno` of a [BGData-class]
+#' object.
+#' @param center_snp Columns of `X` to compute windows. By default, does not attempt to obtain LD windows surrounding
```
I think `range` was a better name for this. `center_snp` makes me think that only a single value is acceptable. Also, the current default value makes the assumption that `X` has more than 100 columns. Should this parameter even have a default value?

agrueneberg commented on R/getWindows.R:
```diff
@@ -0,0 +1,99 @@
+#' @return A list the length of the number of columns of `X` containing sliding LD-based windows. The elements of
+#' the list are named for each "center" SNP used in LD determination. Each element is a vector of SNP names--SNPs
+#' in LD with the "center" SNP.
+#' @export
```
Could you please add an example on how this function is supposed to be used? See https://github.com/QuantGen/BGData/blob/master/R/summarize.R#L23 and https://github.com/QuantGen/BGData/blob/master/man/examples/summarize.R for the example of the summarize() function.

agrueneberg reviewed (commented):
Thanks for the changes. I have had some more ideas, feel free to consider these or ignore them depending on how much time you want to spend on this :)

agrueneberg commented:
Hi Scott, I've added your function to the BGDataExt package, the new staging area for new BGData features: https://github.com/QuantGen/BGDataExt/commit/5471d33d93e26b107a7f6232a9535f2e2d09e498
